### PR TITLE
Fix exhausted local ports preventing new connection (#11936)

### LINKED ISSUE #11936: postgresql HA backend - vault auto seals when it cannot open connection to database due to exhausted local ports
**This issue was caught in QA/stress testing and is not really expected in a production environment**, however could also be forced by users who can login to the vault server host.

Because vault (when using postgresql backend) does fast ephemeral connections to postgresql to create/delete leases and entries in the DB, if there are too many connections in TIME_WAIT or CLOSE_WAIT vault can cycle through the entirety of the port range available for connections and eventually run out of ports, printing the following type of error:

```
Jun 24 19:38:50 vault-2 vault[32594]: 2021-06-24T19:38:50.179+0100 [ERROR] core: failed to create token: error="failed to persist accessor index entry: dial tcp 10.0.0.1:5432: connect: cannot assign requested address"
Jun 24 19:38:50 vault-2 vault[32594]: 2021-06-24T19:38:50.841+0100 [ERROR] core: failed to create token: error="failed to persist accessor index entry: dial tcp 10.0.0.1:5432: connect: cannot assign requested address"
```
after some time of this happening vault errors and seals itself. In the case where vault tries to expire leases upon startup and has too many leases (let's say 200k it needs to expire) vault exhausts ports and then seals itself. This causes vault to become unusable because it keeps on re-sealing itself over and over again.

If the user is persistent and also unseals vault in a loop, then vault will reach a stable point when the number of leases goes below 10k, which can be seen with this query:

```
vault=> select path, count(path) from vault_kv_store group by path having count(path) > 500;
                path                | count
------------------------------------+-------
 /sys/expire/id/auth/approle/login/ |  1826
 /sys/token/accessor/               |  3300
 /sys/token/id/                     |  3968
(3 rows)
```

1. Have a vault cluster (tested on v1.7.2) using the postgresql storage backend. This was tested in a cluster but may also happen on a single vault.
2. Run `vault write auth/approle/login role_id=... secret_id=...` in a loop millions of times until we have 200k+ outstanding vault tokens that are going to expire.
2. stop all vaults in the cluster, let's say due to an upgrade.

3. Make sure you have a large number of outstanding leases in the DB:
```
vault=> select path, count(path) from vault_kv_store group by path having count(path) > 500;
                path                | count
------------------------------------+--------
 /sys/expire/id/auth/approle/login/ | 226473
 /sys/token/accessor/               | 227460
 /sys/token/id/                     | 227691
(3 rows)
```

4. restart and unseal the vaults

**Expected behavior**
5. vault gets unsealed and works normally doing lease deletion in background.

**Observed behaviour**
6. Vault frantically tries to remove expired leases and delete leases from tables, rapidly cycling and exhausting all source ports to the postgresql server. When no ports are available anymore vault starts erroring and then seals itself.

```hcl
cluster_name            = "test"
log_level               = "trace"
pid_file                = "/run/vault_pgsql.pid"

ui                      = true
disable_mlock           = true
verbose_oidc_logging    = true
raw_storage_endpoint    = true

# must have full protocol in the string
cluster_addr      = "https://..."
api_addr          = "https://..."

tls_require_and_verify_client_cert = "false"

listener "tcp" {
  address = "10.0.0.10:9999"
  tls_disable  = "false"
  tls_disable_client_certs = "true"
  tls_cert_file ="/etc/vault.d/tls/vault.crt"
  tls_key_file ="/etc/vault.d/tls/vault.key"


storage "postgresql" {
    connection_url = "postgres://vault_user:password@dbhost:5432/vault?sslmode=disable"
    ha_enabled     = true
}
```

### PR DESCRIPTION
**Pull request by CuriousCorrelation**
Closes #11936 according to https://github.com/hashicorp/vault/issues/11936#issuecomment-876745591

### CODE REVIEW

**Comment by hsimon-hashicorp** [comment]
Hi! Please don't forget to include a changelog entry. :)

File: website/content/docs/configuration/storage/postgresql.mdx [hunk 1]
```diff
@@ -99,7 +99,7 @@ LANGUAGE plpgsql;
   which to write Vault data. This table must already exist (Vault will not
   attempt to create it).
 
-- `max_idle_connections` `(int)` - Default not set. Sets the maximum number of
+- `max_idle_connections` `(string: "256")` - Sets the maximum number of
```
[the remark this answers is not in the document]
Comment (CuriousCorrelation) [reply website/content/docs/configuration/storage/postgresql.mdx:102]: No worries at all! Shall I resolve this convo then?

**Comment by Zlaticanin** [comment]
Hello @CuriousCorrelation, thank you for working on this. Is this still something you are interested in? If so, could you please take a look at the suggestions Max brought up? Thank you!

**Comment by CuriousCorrelation** [comment]
@Zlaticanin Hi! Yes, still interested. Unfortunately, I didn't receive notification for the requested changes. Currently, I have a busy schedule, but I will make time to make the changes soon. However, in the meantime if you need someone else to take over, I completely understand.

**Comment by hsimon-hashicorp** [comment]
@CuriousCorrelation Are you still interested in working on this? If you're still working with a busy schedule, please feel free to let us know and we can see if a Vault engineer can pick it up. Thanks!

**Comment by CuriousCorrelation** [comment]
@hsimon-hashicorp Yes unfortunately I haven't had the opportunity to review this, and it appears unlikely that I'll be able to do so in the future. Please proceed with the reassignment.

**Comment by VioletHynes** [comment]
Hi there! I'm going to close this given the previous comments. This isn't a high priority for us to be picked up internally right now and I don't expect it will be in the future. That being said, if you later want to re-raise this as a new PR with the suggestions implemented, feel free.

Thanks for your contribution nonetheless!

**Pull request closed without merging**